Changes in module 'nf-core/kalign/align'
--- modules/nf-core/kalign/align/main.nf
+++ modules/nf-core/kalign/align/main.nf
@@ -6,6 +6,13 @@
     container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
         'https://depot.galaxyproject.org/singularity/mulled-v2-5cd0277547c6b33133225c8ce14c0cf2a4396ea2:0a70b6d89a3e06fbdc4a735461e8b98ff32ee5de-0':
         'biocontainers/mulled-v2-5cd0277547c6b33133225c8ce14c0cf2a4396ea2:0a70b6d89a3e06fbdc4a735461e8b98ff32ee5de-0' }"
+
+    errorStrategy  {
+        if (task.exitStatus == 132) {
+            log.warn "KALIGN failed because is incompatible with some CPU types - error will be ignored."
+            return 'ignore'
+        }
+    }
 
     input:
     tuple val(meta), path(fasta)

************************************************************
