Changes in module 'nf-core/tcoffee/seqreformat'
--- modules/nf-core/tcoffee/seqreformat/main.nf
+++ modules/nf-core/tcoffee/seqreformat/main.nf
@@ -38,7 +38,7 @@
     def args = task.ext.args ?: ''
     prefix = task.ext.prefix ?: "${meta.id}"
     """
-    touch "${prefix}_${seq_reformat_type}.txt"
+    touch "${prefix}.txt"
 
     cat <<-END_VERSIONS > versions.yml
     "${task.process}":

--- modules/nf-core/tcoffee/seqreformat/environment.yml
+++ /dev/null
@@ -1,7 +0,0 @@
-name: tcoffee_seqreformat
-channels:
-  - conda-forge
-  - bioconda
-  - defaults
-dependencies:
-  - bioconda::t-coffee=13.46.0.919e8c6b

--- modules/nf-core/tcoffee/seqreformat/tests/main.nf.test.snap
+++ /dev/null
@@ -1,23 +0,0 @@
-{
-    "formatted_file": {
-        "content": [
-            [
-                [
-                    {
-                        "id": "test"
-                    },
-                    "test.txt:md5,fcd4691daf120c88ec5def7ac06fb562"
-                ]
-            ]
-        ],
-        "timestamp": "2023-11-28T11:56:22.705185493"
-    },
-    "versions": {
-        "content": [
-            [
-                "versions.yml:md5,68fb841e6e44274d430a1382bb0bbd14"
-            ]
-        ],
-        "timestamp": "2023-11-28T11:56:22.717235196"
-    }
-}
--- modules/nf-core/tcoffee/seqreformat/tests/tags.yml
+++ /dev/null
@@ -1,2 +0,0 @@
-tcoffee/seqreformat:
-  - "modules/nf-core/tcoffee/seqreformat/**"

--- modules/nf-core/tcoffee/seqreformat/tests/nextflow.config
+++ /dev/null
@@ -1,3 +0,0 @@
-process {
-    ext.args = "-output=sim_idscore"
-}

--- modules/nf-core/tcoffee/seqreformat/tests/main.nf.test
+++ /dev/null
@@ -1,33 +0,0 @@
-nextflow_process {
-
-    name "Test Process TCOFFEE_SEQREFORMAT"
-    script "../main.nf"
-    process "TCOFFEE_SEQREFORMAT"
-
-    tag "modules"
-    tag "modules_nfcore"
-    tag "tcoffee"
-    tag "tcoffee/seqreformat"
-
-    test("sarscov2 - bam") {
-
-        when {
-            process {
-                """
-                input[0] = [ [ id:'test' ],
-                             file("https://raw.githubusercontent.com/nf-core/test-datasets/multiplesequencealign/testdata/setoxin-ref.fa", checkIfExists: true)
-                           ]
-                """
-            }
-        }
-
-
-        then {
-            assertAll(
-                { assert process.success },
-                { assert snapshot(process.out.formatted_file).match("formatted_file")},
-                { assert snapshot(process.out.versions).match("versions") }
-            )
-        }
-    }
-}

************************************************************
